compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java

```diff
@@ -661,14 +661,30 @@ private boolean notInSameScope(BLangVariableReference varRefExpr, SymbolEnv env)
     }
 
     private boolean isMarkedTainted(BLangVariableReference varRef) {
+        if (isStructuredAccessOnVariableReference(varRef)) {
+            return isMarkedTainted((BLangVariableReference) ((BLangAccessExpression) varRef).expr);
+        }
         return ((BVarSymbol) varRef.symbol).taintabilityAllowance == BVarSymbol.TaintabilityAllowance.TAINTED;
     }
 
+    private boolean isStructuredAccessOnVariableReference(BLangVariableReference variableReference) {
+        return (variableReference.getKind() == NodeKind.INDEX_BASED_ACCESS_EXPR
+                || variableReference.getKind() == NodeKind.FIELD_BASED_ACCESS_EXPR
+                || variableReference.getKind() == NodeKind.XML_ATTRIBUTE_ACCESS_EXPR)
+                && ((BLangAccessExpression) variableReference).expr instanceof BLangVariableReference;
```

Comment:
Do we need the check for `((BLangAccessExpression) variableReference).expr instanceof BLangVariableReference`?

Reply:
I was under wrong impression about BLangVariableReference. Will remove it